### PR TITLE
Joy demystifying regex

### PR DESCRIPTION
**Pull request by pm0kjp**
This is module 1 of a three-module series.

### CODE REVIEW

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+Other rules might describe the pattern of a valid credit card number, or a national identity number, or a postal address.  
+
+As humans, we are very good at identifying patterns and realizing instantly that a certain text "looks like an email address" or "looks like a phone number".  Quantifying patterns and telling the computer what a pattern "looks like" is what regular expressions are all about.
+
+<h3>Patterns in Biomedical Research</h3>
+
+In biomedical research, patterns emerge in a number of ways:
+
+* File names might follow a pattern that include a subject identifier and date stamp, which can be useful in filtering out only file names from a certain time period or for a certain subject.
+* Clinical notes in an electronic health record might contain text like " R/O ", " RO:", " r/o " which mean "rule out" or "ruled out", which are important to find in order to discover differential diagnosis clues.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:60]: This sentence is all of my grammar/comma pet peeves, feel free to just say "stop pestering me about this it isn't important" but here is a suggestion to "fix" the comma problems... or at least give me less anxiety about commas going inside or outside of quotes.

```suggestion
* Clinical notes in an electronic health record might represent the phrases "rule out" or "ruled out" as `R/O`, `RO:`,  or `r/o`. It is important to find all of these in order to discover differential diagnosis clues.
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+We like "Regex 101", at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.  
+
+Below, or by [trying it out yourself](https://regex101.com/r/QUok4D), you can see that we put in our "valid URL" regular expression (at the top of the screen, in the appropriately labeled box) as well as the different varieties of URLs, some of which we think are valid and match our pattern, and some of which we think are invalid and should fail to match our pattern.
+
+If our test strings match the pattern, they are highlighted, and if they do not match, they are not highlighted.
+
+![Regex 101 application with sample URL regex pattern and test URL strings.](media/url_regex.png)<!-- style = "border: 1px solid rgb(var(--color-highlight));" -->
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:220]: Weird behavior in safari browser: it allowed me to click on some of the links in the image to navigate to the websites. This led me to investigate whether this was an iframe or a normal image. This does not occur in Chrome so I don't think it is a LiaScript thing, just some weird Safari "feature" that feels kind of like a bug to me in this setting.

I don't have a suggestion for changing anything, I just wanted to record this behavior somewhere.

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+
+## Quiz: Demystifying Regex
+
+
+Fill in the blank below with a single lowercase word.
+---
+
+Regular expressions are a way to define rules that describe a 
+
+[[ pattern ]] in text.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:232]: I feel like the this question needs a little restructuring. I'm not sure coming up with the word "pattern" is the most useful way to meet the learning objective of explaining what a regular expression is. I need to ponder what type of questions might get at this learning objective better and get back to you.

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+
+## Regular Expressions, Defined
+
+The word *regular* (like *regulate*) comes from the Latin root for "rule": *regula*.  A **regular expression** is a specific way to **express** a **rule** for a pattern.  
+
+For example, one pattern rule might be for the URL of a root web page, like https://www.nytimes.com or http://apache.org.  
+
+Other rules might describe the pattern of a valid credit card number, or a national identity number, or a postal address.  
+
+As humans, we are very good at identifying patterns and realizing instantly that a certain text "looks like an email address" or "looks like a phone number".  Quantifying patterns and telling the computer what a pattern "looks like" is what regular expressions are all about.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:53]:
```suggestion
As humans, we are very good at identifying patterns and realizing instantly that a certain text "looks like an email address" or "looks like a phone number."  Quantifying patterns and telling the computer what a pattern "looks like" is what regular expressions are all about.
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+
+<h3>Patterns in Biomedical Research</h3>
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:55]: Maybe make this an actual subsection, and move the goal of regular expressions from the bottom of the next page to its own subsection here as well?

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+
+Regex is often used in **code**.  
+
+For example, you might write Python code that searches for patterns that look like medical record numbers and redacts them, replacing them with "9999999".  
+
+Or you could write R code that takes the file names from a directory full of video files with specially formatted file names like `subj412834_20230412_ALC.mp4` and makes a table of the information those file names contain (like subject id, date, initials of who made the interview recording, etc.).
+
+Regular expressions can also be used in **text editors**.  
+
+For example, this module was written using a text editor called [Visual Studio Code](https://code.visualstudio.com/) (or VS Code).  This text editor allows the user to search for patterns using regular expressions.  For example, the author can look for every time a word appears between single asterisks (which will mean the text will appear in italics in the module when a learner reads **).
```
[the remark this answers is not in the document]
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:118]: I .... also don't know what I was attempting to communicate?  lol I will edit

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 4 above)
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:53]: I love that you have an eye for this stuff and it's my most frequent mistake!

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+
+Regular expressions can also be used in **text editors**.  
+
+For example, this module was written using a text editor called [Visual Studio Code](https://code.visualstudio.com/) (or VS Code).  This text editor allows the user to search for patterns using regular expressions.  For example, the author can look for every time a word appears between single asterisks (which will mean the text will appear in italics in the module when a learner reads **).  
+
+VS Code also supports search and replace.  For example, if there's a style rule that indicates that words in *italics* (like `*regular*`) should also be in bold, the author can ask the text editor to find all the places where there's a word enclosed in single asterisks and enclose that word instead in triple asterisks (like `***regular***`, which will make the word render in ***bold italic***).
+
+There are also some popular **command line tools** that use regular expressions, like `sed`, `grep`, and `awk`.
+
+## An Example Regular Expression
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:124]: Structural suggestion: 
There is really one big example here, with 3 components:
1. breaking down the pattern of what "looks like" a website
2. parsing the code for that pattern in regex
3. showing that that pattern does match the things we want it to match and doesn't match things we don't.

I suggest reordering it this way, but if this makes less sense to you, please push back.

---

## An Example Regular Expression
Hey, let's investigate a regular expression for the pattern that makes up a url. We want it to match these [list] but not these [list].
### Describe the pattern in words
current contents of lines 130-146, describing in English what the pattern is
### There exists code that describes the same pattern in words
Introduce Regex 101. If you asked someone who knows regex to write down your pattern, this is what they would give you, and you can check that it works
### How does the code encode the pattern?
Content from current lines 178-207

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 3 above)
Comment (drelliche) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:232]: Ok, I think I'd like this better as a "Check all that are true" question with options like:
- [[X]] Regular expressions are a way to define rules that describe a pattern in text.
- [[X]] Regular expressions can be used in multiple coding languages.
- [[X]] Regular expressions can be used to extract useful substrings.
- [[ ]]Regular expressions are easy to learn and simple to memorize.

**Review by drelliche — COMMENTED** [review]
Overall looks great! I have some nitpick-y comma things and a few structural suggestions for moving chunks around.

I'm not 100% sure my suggestions improve clarity in general, they might just be closer to how I think, so feel free to reject those suggestions, but take a look and then I can give it one more careful comb through before approving :)

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 2 above)
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:220]: WHOA!  that's weird and maybe good?  IDK

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 5 above)
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:55]: done in new commit!

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:120]: OK, fixed, and added something just for you:

> For example, the author can look for every time a word appears between single asterisks (which will mean the text will appear in italics), or every time a new line starts with a number, or every time a single or double quotation mark is followed by a period or comma.

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+
+@learning_objectives  
+After completion of this module, learners will be able to:
+
+- Explain what a regular expression is 
+- Give an example of how regular expressions can be useful
+- Use an online regular expressions checker that helps build and test regular expressions.
+
+@end
+
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:33]: Metadata to mesh with docs.md:

```suggestion
good_first_module: false
coding_required: true
coding_level: getting_started
sequence_name: regex
previous_sequential_module: 
@sets_you_up_for
- regular_expressions_basics
@end 
@depends_on_knowledge_available_in
@end
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 3 above)
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:234]: I agree!

File: demystifying_regular_expressions/demystifying_regular_expressions.md (hunk 7 above)
Comment (pm0kjp) [reply demystifying_regular_expressions/demystifying_regular_expressions.md:128]: updated in commit!

**Comment by pm0kjp** [comment]
@drelliche okay I added some changes that I think meet your suggestions!  Thank you!  Give it a once over if you have a chance.

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+You can also put in a "capturing" element, to capture a specific part of the pattern (for example, extracting the patient identifier and/or the sample number from a file name).
+
+Some regular expressions are short and simple, and some can be many lines of text long.
+
+## Why Use Regular Expressions?
+
+There are two very common reasons you as a researcher might want to use regular expressions (also referred to as "regex", which you can pronounce either with "reg" rhyming with "beg" or "reg" rhyming with "wedge").
+
+**Reason 1: Searching for the needle in the haystack
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:100]:
```suggestion
**Reason 1: Searching for the needle in the haystack**
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+Regular expressions require a bit of effort to learn, and even people who use regular expressions often will end up making mistakes and using online tools for help.  
+
+If you work with strings (character data) frequently, whether that's from free-text entries in a REDCap database, text from web-scraped blogs, file names from a huge dataset, or clinical notes from an electronic health record, the learning curve is definitely worth it!  
+
+</div>
+
+## Where Can You Use Regular Expressions?
+
+Regex is often used in **code**.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:123]:
```suggestion
**Regex is often used in **code**.**
---
```
Just a suggestion to visually match the page before it :)

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+## Where Can You Use Regular Expressions?
+
+Regex is often used in **code**.  
+
+For example, you might write Python code that searches for patterns that look like medical record numbers and redacts them, replacing them with "9999999".  
+
+Or you could write R code that takes the file names from a directory full of video files with specially formatted file names like `subj412834_20230412_ALC.mp4` and makes a table of the information those file names contain (like subject id, date, initials of who made the interview recording, etc.).
+
+Regular expressions can also be used in **text editors**.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:129]:
```suggestion
Regular expressions can also be used in **text editors**.  
---
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+
+Regular expressions can also be used in **text editors**.  
+
+For example, this module was written using a text editor called [Visual Studio Code](https://code.visualstudio.com/) (or VS Code).  This text editor allows the user to search for patterns using regular expressions.  For example, the author can look for every time a word appears between single asterisks (which will mean the text will appear in italics), or every time a new line starts with a number, or every time a single or double quotation mark is followed by a period or comma.
+
+VS Code also supports search and replace.  For example, if there's a style rule that indicates that words in *italics* (like `*regular*`) should also be in bold, the author can ask the text editor to find all the places where there's a word enclosed in single asterisks and enclose that word instead in triple asterisks (like `***regular***`, which will make the word render in ***bold italic***).
+
+There are also some popular **command line tools** that use regular expressions, like `sed`, `grep`, and `awk`.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:135]:
```suggestion
There are also some popular **command line tools** that use regular expressions.
---

These include commands like `sed`, `grep`, and `awk`.
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 13]
```diff
@@ -0,0 +1,331 @@
...
+
+If you work with strings (character data) frequently, whether that's from free-text entries in a REDCap database, text from web-scraped blogs, file names from a huge dataset, or clinical notes from an electronic health record, the learning curve is definitely worth it!  
+
+</div>
+
+## Where Can You Use Regular Expressions?
+
+Regex is often used in **code**.  
+
+For example, you might write Python code that searches for patterns that look like medical record numbers and redacts them, replacing them with "9999999".
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:125]:
```suggestion
For example, you might write Python code that searches for patterns that look like medical record numbers and redacts them, replacing them with 9999999.  
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 14]
```diff
@@ -0,0 +1,331 @@
...
+* `\.` : A literal period (`.`)
+* `[A-Za-z0-9\-\.]+`: One or more characters that's either in the A-Z range, the a-z range, the 0-9 range, a literal hyphen (`-`), or a literal period (`.`)
+* `\/?`: Optionally, a final forward slash
+
+
+## A Great Tool to Start With
+
+As a final step in demystifying regular expressions, it can be helpful to introduce the online regex checker as a great resource to learn and to check your regex before trying to learn how to make your own regular expressions or use regex in SQL, R, or Python code you're writing.
+
+We like "Regex 101", at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:238]:
```suggestion
We like **Regex 101**, at https://www.regex101.com, but you can also search online for "regex checker" to find other, similar websites.  Regex 101 allows you to enter your regular expression as well as some sample text that you want to apply the regular expression to.  
```

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+[[ ]] Regular expressions are a way to detect statistical outliers in data.
+[[X]] Regular expressions are a way to define rules that describe a pattern in text.
+[[X]] Regular expressions can be used in multiple coding languages.
+[[ ]] The only way to check the accuracy of a regular expression someone gives you is by writing code. 
+[[X]] Regular expressions can be used to extract useful substrings.
+
+
+***
+<div class = "answer">
+
```
Comment (drelliche) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:261]: This div box is rendering incorrectly... I think removing the white space between the last option and the `***` might fix it?

**Review by drelliche — COMMENTED** [review]
@paytonk A few tiny new things and then it is ready!

File: demystifying_regular_expressions/demystifying_regular_expressions.md [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+Please check all of the true statements below:
+
+[[ ]] Regular expressions are a way to detect statistical outliers in data.
+[[X]] Regular expressions are a way to define rules that describe a pattern in text.
+[[X]] Regular expressions can be used in multiple coding languages.
+[[ ]] The only way to check the accuracy of a regular expression someone gives you is by writing code. 
+[[X]] Regular expressions can be used to extract useful substrings.
+
+
+***
```
Comment (pm0kjp) [inline demystifying_regular_expressions/demystifying_regular_expressions.md:259]:
```suggestion
[[X]] Regular expressions can be used to extract useful substrings.
***
```

**Comment by pm0kjp** [comment]
@drelliche stick a fork in it, it's done!

**Review by drelliche — APPROVED** [review]
LGTM!

**Pull request merged**